url3472127e1f_SergeyKrivohatskiy_tiny_database_tgz-pJ8-ru_spbau_tinydb_bufferManager_IndexTestJ8/src/ru/spbau/tinydb/bufferManager/BufferManager.java

```diff
@@ -0,0 +1,160 @@
+    private int currentPos = -1;
+    private int length;
+
+    public BufferManager(String dbFileName) throws DBException {
+        this(new File(dbFileName));
+    }
+
+    public BufferManager(File dbFile) throws DBException {
+        try {
+            this.dbFile = new RandomAccessFile(dbFile, "rw");
+        } catch (FileNotFoundException e) {
+            throw new DBException(e);
+        }
+
+        init();
+    }
+
+    private void init() throws DBException {
+        Arrays.fill(bufferPosToPageIdx, -1);
+        try {
+            length = (int) (dbFile.length() / PAGE_SIZE);
+            if(length == 0) {
+                increaseLength();
+                //writeHeader(there is no header now)
+                increaseLength(); // for MetaInf Page
+                increaseLength(); // for Indexes Page
+            } else {
+                //loadHeader(there is no header now)
+            }
+        } catch (IOException e) {
+            throw new DBException("IOException in BufferManager init method(getting file lenght)");
+        }
+    }
+
+    private void increaseLength() {
+        length += 1;
+        try {
+            dbFile.setLength(length * PAGE_SIZE);
+        } catch (IOException e) {
+            throw new RuntimeException("IOException when increasing file size. No space left");
+        }
+    }
+
+    public int getFreePage() {
+        increaseLength();
+        return length - 1;
+    }
+
+    public BufferView getPage(int pageIndex) {
+        if(pageIdxToBufferPos.containsKey(pageIndex)) {
+            int bufferPos = pageIdxToBufferPos.get(pageIndex);
+            return new BufferView(buffer, bufferPos, (int) PAGE_SIZE, this);
+        }
+        
+        // Clock algorithm
+        while(true) {
+	        currentPos = bufferIsUsed.nextClearBit(currentPos + 1);
+	        if(currentPos == BUFFER_SIZE) {
+	        	currentPos = bufferIsUsed.nextClearBit(0);
+	        }
+	        if(currentPos == BUFFER_SIZE) {
+	            throw new RuntimeException("All buffer positions are in use! Panic!");
+	        }
+	        if(bufferWasUsed.get(currentPos)) {
+	        	bufferWasUsed.clear(currentPos);
+	        } else {
+	        	break;
+	        }
+        }
+        if(bufferChanged.get(currentPos)) {
+            writePage(currentPos);
+        }
+        pageIdxToBufferPos.remove(bufferPosToPageIdx[currentPos]);
+        bufferPosToPageIdx[currentPos] = -1;
+        loadPage(pageIndex, currentPos);
+        return new BufferView(buffer, currentPos, (int) PAGE_SIZE, this);
+    }
+
+    private void writePage(int bufferPos) {
+        try {
+            int pageIndex = bufferPosToPageIdx[bufferPos];
+            dbFile.seek(pageIndex * PAGE_SIZE);
+            dbFile.write(buffer.array(), (int) (bufferPos * PAGE_SIZE), (int) PAGE_SIZE);
+            bufferChanged.clear(bufferPos);
+        } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)